core/server/api_container/server/startosis_engine/kurtosis_instruction/run_sh/run_sh.go

```diff
@@ -0,0 +1,237 @@
+func (builtin *RunShCapabilities) Execute(ctx context.Context, _ *builtin_argument.ArgumentValuesSet) (string, error) {
+	// create work directory and cd into that directory
+	createAndSwitchTheDirectory := fmt.Sprintf("mkdir -p %v && cd %v", builtin.workdir, builtin.workdir)
+
+	// replace future references to actual strings
+	maybeSubCommandWithRuntimeValues, err := magic_string_helper.ReplaceRuntimeValueInString(builtin.run, builtin.runtimeValueStore)
+	if err != nil {
+		return "", stacktrace.Propagate(err, "An error occurred while replacing runtime values in run_sh")
+	}
+
+	commandWithNoNewLines := strings.ReplaceAll(maybeSubCommandWithRuntimeValues, "\n", " ")
+
+	completeRunCommand := fmt.Sprintf("%v && %v", createAndSwitchTheDirectory, commandWithNoNewLines)
+	createDefaultDirectory := []string{"/bin/sh", "-c", completeRunCommand}
+	serviceConfigBuilder := services.NewServiceConfigBuilder(builtin.image)
+	serviceConfigBuilder.WithFilesArtifactMountDirpaths(builtin.files)
+	// This make sure that the container does not stop as soon as it starts
+	// This only is needed for kubernetes at the moment
+	// TODO: Instead of creating a service and running exec commands
+	//  we could probably run the command as an entrypoint and retrieve the results as soon as the
+	//  command is completed
+	serviceConfigBuilder.WithCmdArgs([]string{"tail", "-f", "/dev/null"})
+
+	serviceConfig := serviceConfigBuilder.Build()
+	_, err = builtin.serviceNetwork.AddService(ctx, service.ServiceName(builtin.name), serviceConfig)
+
+	if err != nil {
+		return "", err
+	}
+
+	// run the command passed in by user in the container
+	code, output, err := builtin.serviceNetwork.ExecCommand(ctx, builtin.name, createDefaultDirectory)
+	if err != nil {
+		return "", err
+	}
+
+	result := map[string]starlark.Comparable{
+		runshOutputKey: starlark.String(output),
+		runshCodeKey:   starlark.MakeInt(int(code)),
+	}
+	builtin.runtimeValueStore.SetValue(builtin.resultUuid, result)
+
+	instructionResult := resultMapToString(result)
+	return instructionResult, err
+}
+
+// Copied some of the command from: exec_recipe.ResultMapToString
+// TODO: create a utility method that can be used by add_service(s) and run_sh method.
+func resultMapToString(resultMap map[string]starlark.Comparable) string {
+	exitCode := resultMap[runshCodeKey]
+	rawOutput := resultMap[runshOutputKey]
+	outputStarlarkStr, ok := rawOutput.(starlark.String)
+	if !ok {
+		logrus.Errorf("Result of an exec recipe was not a string (was: '%v' of type '%s'). This is not fatal but the object might be malformed in CLI output. It is very unexpected and hides a Kurtosis internal bug. This issue should be reported", rawOutput, reflect.TypeOf(rawOutput))
+		outputStarlarkStr = starlark.String(outputStarlarkStr.String())
+	}
+	outputStr := outputStarlarkStr.GoString()
+	if outputStr == "" {
+		return fmt.Sprintf("Command returned with exit code '%v' with no output", exitCode)
```

Comment:
isn't worth it to check if exitCode is an int or not?

Reply:
I am keeping it consistent to `exec` -- though good point; we probably can change it exec as well.